## Intro

I want to implement wav2vec2 from scratch to better understand it. I use toy dataset with few audio files. I want to see how pretraining and finetuning works. I want to have dataset, dataloader, feature extractor, quantization, transformer modules

what plan do you suggest to use?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great project! Implementing wav2vec2 from scratch is an excellent way to understand self-supervised speech learning. Here's a plan I'd suggest:

**Phase 1: Setup & Data Pipeline**
- Load audio files and create a simple dataset class
- Build a dataloader that handles batching
- Create basic audio preprocessing (resampling, normalization)

**Phase 2: Core Components**
- Feature extractor (CNN layers to convert raw audio to latent representations)
- Quantization module (vector quantization for creating discrete targets)
- Transformer encoder (context network)

**Phase 3: Pretraining**
- Contrastive loss implementation
- Masking strategy for the input
- Training loop for self-supervised learning

**Phase 4: Finetuning** 
- Add a classification head for your specific task
- Finetuning loop with labeled data

Since you're using a toy dataset, we can simplify some aspects (fewer layers, smaller dimensions) to make training feasible.

What's your experience level with PyTorch and transformers? And do you have a specific downstream task in mind for the finetuning phase (like speech recognition, speaker classification, etc.)?

I have used transformers and pytorch. I think about automatic speech recognition

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! ASR is a great choice for finetuning. Since you're familiar with PyTorch and transformers, we can move at a good pace.

Let's start with Phase 1. For the dataset, you'll need:
- A way to load your audio files (librosa or torchaudio work well)
- A Dataset class that returns raw waveforms
- A DataLoader with collation to handle variable-length audio

Do you already have your toy audio dataset ready? And would you like to start by creating the Dataset class yourself, or would you like me to show you a basic structure first?

## Dataset

lets use tochaudio and I have one audio file "emt16k.wav"

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great! Let's start with torchaudio and your single audio file.

First, you'll want to load the audio file to see what you're working with. Try loading it and checking its properties:

```python
import torchaudio
waveform, sample_rate = torchaudio.load("emt16k.wav")
```

Once you load it, check the shape of the waveform and the sample rate. This will tell us if it's mono/stereo and how long it is.

Have you tried loading the file yet? What do you see?

In [ ]:
import torchaudio
waveform, sample_rate = torchaudio.load("emt16k.wav")
waveform.shape, sample_rate

(torch.Size([1, 203815]), 16000)

this, please give me dataset class code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's a basic Dataset class for your audio file:

```python
import torch
from torch.utils.data import Dataset

class AudioDataset(Dataset):
    def __init__(self, audio_files, sample_rate=16000):
        self.audio_files = audio_files
        self.sample_rate = sample_rate
    
    def __len__(self):
        return len(self.audio_files)
    
    def __getitem__(self, idx):
        waveform, sr = torchaudio.load(self.audio_files[idx])
        # Resample if needed
        if sr != self.sample_rate:
            resampler = torchaudio.transforms.Resample(sr, self.sample_rate)
            waveform = resampler(waveform)
        # Convert to mono if stereo
        if waveform.shape[0] > 1:
            waveform = torch.mean(waveform, dim=0, keepdim=True)
        return waveform.squeeze(0)
```

You can create an instance like:
```python
dataset = AudioDataset(["emt16k.wav"])
```

Since you have just one file, for pretraining you might want to split it into chunks. Should we add chunking logic to create multiple training samples from your single audio file?

In [ ]:
import torch
from torch.utils.data import Dataset

class AudioDataset(Dataset):
    def __init__(self, audio_files, sample_rate=16000):
        self.audio_files = audio_files
        self.sample_rate = sample_rate
    
    def __len__(self):
        return len(self.audio_files)
    
    def __getitem__(self, idx):
        waveform, sr = torchaudio.load(self.audio_files[idx])
        # Resample if needed
        if sr != self.sample_rate:
            resampler = torchaudio.transforms.Resample(sr, self.sample_rate)
            waveform = resampler(waveform)
        # Convert to mono if stereo
        if waveform.shape[0] > 1:
            waveform = torch.mean(waveform, dim=0, keepdim=True)
        return waveform.squeeze(0)

In [ ]:
dataset = AudioDataset(["emt16k.wav","emt16k.wav","emt16k.wav"])
dataset

now to dataloader?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Before the dataloader, we need to handle the chunking issue. Your audio file is 203,815 samples long (about 12.7 seconds), but for training you'll need multiple samples.

You have a few options:

1. **Chunk in the Dataset** - Split the audio into fixed-length segments (e.g., 3 seconds each) in `__getitem__`
2. **Chunk beforehand** - Pre-process the audio into separate chunks
3. **Use the whole file** - Keep it as one sample (less ideal for learning)

For wav2vec2 pretraining, option 1 is most common. You'd modify `__getitem__` to return random or sequential chunks of a fixed length.

Which approach would you like to try?

lets reuse this file multiple times. So I load this file 3 times now

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Got it! You're loading the same file 3 times to have more samples. That works for a toy example.

Now for the DataLoader, you'll need a collate function to handle batching, since audio samples might have different lengths (though yours are currently identical).

Here's a simple collate function:

```python
def collate_fn(batch):
    # Pad sequences to same length
    lengths = [x.shape[0] for x in batch]
    max_len = max(lengths)
    padded = torch.stack([torch.nn.functional.pad(x, (0, max_len - x.shape[0])) for x in batch])
    return padded, torch.tensor(lengths)
```

Try creating a DataLoader using this collate function. What batch size do you want to use?

In [ ]:
def collate_fn(batch):
    # Pad sequences to same length
    lengths = [x.shape[0] for x in batch]
    max_len = max(lengths)
    padded = torch.stack([torch.nn.functional.pad(x, (0, max_len - x.shape[0])) for x in batch])
    return padded, torch.tensor(lengths)

provide dataloader class

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

You don't need a custom DataLoader class - just use PyTorch's built-in one:

```python
from torch.utils.data import DataLoader

dataloader = DataLoader(dataset, batch_size=2, collate_fn=collate_fn, shuffle=True)
```

Try creating it and then iterate through one batch to see what you get. What happens when you do:

```python
batch = next(iter(dataloader))
```

In [ ]:
from torch.utils.data import DataLoader

dataloader = DataLoader(dataset, batch_size=2, collate_fn=collate_fn, shuffle=True)

In [ ]:
batch = next(iter(dataloader))

In [ ]:
batch

(tensor([[ 0.0073,  0.0147,  0.0128,  ..., -0.0063, -0.0076, -0.0140],
         [ 0.0073,  0.0147,  0.0128,  ..., -0.0063, -0.0076, -0.0140]]),
 tensor([203815, 203815]))

what next?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! Your dataloader is working - you have batches of waveforms and their lengths.

Now let's move to **Phase 2: Core Components**. The first component is the **feature extractor** - a CNN that converts raw audio waveforms into latent representations.

In wav2vec2, this is typically a stack of 1D convolutional layers with specific strides. For a toy version, we could use 4-5 conv layers.

Would you like to:
1. Try building the feature extractor CNN yourself based on a description?
2. Have me provide the code structure for you to modify?

What's your preference?

## Feature extraction

provide code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the feature extractor CNN:

```python
import torch.nn as nn

class FeatureExtractor(nn.Module):
    def __init__(self, conv_layers=[(512, 10, 5)] + [(512, 3, 2)] * 4 + [(512, 2, 2)] * 2):
        super().__init__()
        # conv_layers: list of (out_channels, kernel_size, stride)
        layers = []
        in_channels = 1
        for out_channels, kernel_size, stride in conv_layers:
            layers.extend([
                nn.Conv1d(in_channels, out_channels, kernel_size, stride, bias=False),
                nn.GroupNorm(out_channels, out_channels),
                nn.GELU()
            ])
            in_channels = out_channels
        self.conv = nn.Sequential(*layers)
    
    def forward(self, x):
        # x: (batch, time)
        x = x.unsqueeze(1)  # (batch, 1, time)
        x = self.conv(x)  # (batch, channels, time')
        return x.transpose(1, 2)  # (batch, time', channels)
```

Try instantiating it and passing your batch through to see the output shape. What dimensions do you get?

In [ ]:
import torch.nn as nn

class FeatureExtractor(nn.Module):
    def __init__(self, conv_layers=[(512, 10, 5)] + [(512, 3, 2)] * 4 + [(512, 2, 2)] * 2):
        super().__init__()
        # conv_layers: list of (out_channels, kernel_size, stride)
        layers = []
        in_channels = 1
        for out_channels, kernel_size, stride in conv_layers:
            layers.extend([
                nn.Conv1d(in_channels, out_channels, kernel_size, stride, bias=False),
                nn.GroupNorm(out_channels, out_channels),
                nn.GELU()
            ])
            in_channels = out_channels
        self.conv = nn.Sequential(*layers)
    
    def forward(self, x):
        # x: (batch, time)
        x = x.unsqueeze(1)  # (batch, 1, time)
        x = self.conv(x)  # (batch, channels, time')
        return x.transpose(1, 2)  # (batch, time', channels)

In [ ]:
feature_extractor=FeatureExtractor()
features=feature_extractor(batch[0])
features.shape

torch.Size([2, 636, 512])

like so?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Exactly! Your feature extractor is working perfectly. It's taking raw audio waveforms (batch_size=2, time=203815) and converting them to latent features (batch_size=2, time=636, features=512).

Notice how the time dimension went from 203,815 samples down to 636 frames - that's the CNN downsampling the audio by about 320x.

Next up is the **quantization module**. This creates discrete "codebook" representations that will serve as targets during pretraining. It uses vector quantization (VQ) to map continuous features to discrete codes.

Should I provide the quantization module code?

## Quantization

yes, provide

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the quantization module:

```python
class VectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        # Codebook: (num_groups, num_vars, dim // num_groups)
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x):
        # x: (batch, time, dim)
        batch, time, dim = x.shape
        
        # Reshape to groups: (batch, time, num_groups, dim // num_groups)
        x = x.view(batch, time, self.num_groups, -1)
        
        # Compute distances to codebook entries
        distances = torch.sum(x**2, dim=-1, keepdim=True) + \
                   torch.sum(self.codebook**2, dim=-1) - \
                   2 * torch.einsum('btgd,gvd->btgv', x, self.codebook)
        
        # Get nearest codebook entry
        indices = distances.argmin(dim=-1)  # (batch, time, num_groups)
        
        # Quantize
        quantized = torch.stack([self.codebook[g, indices[:,:,g]] for g in range(self.num_groups)], dim=2)
        quantized = quantized.view(batch, time, dim)
        
        return quantized, indices
```

Try creating it and passing your features through. What shapes do you get for the quantized output and indices?

In [ ]:
class VectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        # Codebook: (num_groups, num_vars, dim // num_groups)
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x):
        # x: (batch, time, dim)
        batch, time, dim = x.shape
        
        # Reshape to groups: (batch, time, num_groups, dim // num_groups)
        x = x.view(batch, time, self.num_groups, -1)
        
        # Compute distances to codebook entries
        distances = torch.sum(x**2, dim=-1, keepdim=True) + \
                   torch.sum(self.codebook**2, dim=-1) - \
                   2 * torch.einsum('btgd,gvd->btgv', x, self.codebook)
        
        # Get nearest codebook entry
        indices = distances.argmin(dim=-1)  # (batch, time, num_groups)
        
        # Quantize
        quantized = torch.stack([self.codebook[g, indices[:,:,g]] for g in range(self.num_groups)], dim=2)
        quantized = quantized.view(batch, time, dim)
        
        return quantized, indices

it uses cosine distance?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

No, it's using **Euclidean distance** (L2 distance). The formula:

```
distances = ||x||² + ||codebook||² - 2⟨x, codebook⟩
```

is the expanded form of `||x - codebook||²`.

Wav2vec2 can use either Euclidean or cosine similarity for quantization. If you wanted **cosine similarity** instead, you'd normalize the vectors first and then just compute the dot product (or skip normalization and use negative dot product as the distance).

Would you like to keep Euclidean distance, or switch to cosine similarity?

is tehre a difference?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, there are practical differences:

**Euclidean distance:**
- Sensitive to magnitude - vectors with larger norms are "farther apart"
- Can lead to some codebook entries being used more than others
- More common in general VQ applications

**Cosine similarity:**
- Only cares about direction, not magnitude
- Often leads to more balanced codebook usage
- Used in some wav2vec2 variants

For a toy implementation, either works fine. The original wav2vec2 paper uses a gumbel softmax approach which is closer to cosine similarity behavior.

Do you want to stick with Euclidean (what you have now), or would you like to try cosine similarity?

maybe we should use gumbel softmax?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good idea! Gumbel softmax is actually what the original wav2vec2 uses. It's a differentiable way to sample from the codebook, which helps with training.

With Gumbel softmax, instead of hard assignment to the nearest codebook entry, you get a soft weighted combination during training (with a temperature parameter), but it approaches hard assignment as temperature decreases.

Would you like me to provide a Gumbel softmax version of the quantizer, or would you prefer to keep the current simpler version for now and move on to the transformer?

provide 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the Gumbel softmax quantizer:

```python
import torch.nn.functional as F

class GumbelVectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512, temp=(2.0, 0.5, 0.999995)):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        # Temperature annealing: (start, end, decay)
        self.max_temp, self.min_temp, self.temp_decay = temp
        self.curr_temp = self.max_temp
        
        # Codebook
        self.codebook = nn.Parameter(torch.randn(1, num_groups * num_vars, dim // num_groups))
        
    def forward(self, x, training=True):
        # x: (batch, time, dim)
        batch, time, dim = x.shape
        
        # Reshape: (batch * time * num_groups, dim // num_groups)
        x = x.view(-1, self.num_groups, dim // self.num_groups)
        
        # Compute similarities (negative distances)
        sim = -torch.cdist(x, self.codebook.expand(x.shape[0], -1, -1))
        sim = sim.view(batch * time * self.num_groups, self.num_vars)
        
        if training:
            # Gumbel softmax sampling
            probs = F.gumbel_softmax(sim, tau=self.curr_temp, hard=True, dim=-1)
        else:
            # Hard assignment
            indices = sim.argmax(dim=-1)
            probs = F.one_hot(indices, self.num_vars).float()
        
        # Get quantized vectors
        quantized = probs @ self.codebook.squeeze(0).view(self.num_groups, self.num_vars, -1).view(-1, dim // self.num_groups)
        quantized = quantized.view(batch, time, dim)
        
        return quantized, probs.view(batch, time, self.num_groups, self.num_vars)
```

Try it out and see what shapes you get. You'll also want to add temperature annealing during training (decreasing `curr_temp` over steps).

In [ ]:
import torch.nn.functional as F

class GumbelVectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512, temp=(2.0, 0.5, 0.999995)):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        self.max_temp, self.min_temp, self.temp_decay = temp
        self.curr_temp = self.max_temp
        
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x, training=True):
        batch, time, dim = x.shape
        x = x.view(batch, time, self.num_groups, -1)
        
        # Normalize for cosine similarity
        x_norm = F.normalize(x, dim=-1)
        codebook_norm = F.normalize(self.codebook, dim=-1)
        
        # Compute similarity
        sim = torch.einsum('btgd,gvd->btgv', x_norm, codebook_norm)
        sim = sim.reshape(batch * time * self.num_groups, self.num_vars)
        
        if training:
            probs = F.gumbel_softmax(sim / self.curr_temp, hard=True, dim=-1)
        else:
            indices = sim.argmax(dim=-1)
            probs = F.one_hot(indices, self.num_vars).float()
        
        probs = probs.view(batch * time, self.num_groups, self.num_vars)
        quantized = torch.einsum('bgv,gvd->bgd', probs, self.codebook)
        #quantized = quantized.view(batch, time, dim)
        quantized = quantized.reshape(batch, time, dim)
        
        return quantized, probs.view(batch, time, self.num_groups, self.num_vars)


lets use cosine

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the cosine similarity version:

```python
class VectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x):
        batch, time, dim = x.shape
        x = x.view(batch, time, self.num_groups, -1)
        
        # Normalize for cosine similarity
        x_norm = torch.nn.functional.normalize(x, dim=-1)
        codebook_norm = torch.nn.functional.normalize(self.codebook, dim=-1)
        
        # Compute cosine similarity (use negative for "distance")
        similarity = torch.einsum('btgd,gvd->btgv', x_norm, codebook_norm)
        
        indices = similarity.argmax(dim=-1)  # (batch, time, num_groups)
        
        quantized = torch.stack([self.codebook[g, indices[:,:,g]] for g in range(self.num_groups)], dim=2)
        quantized = quantized.view(batch, time, dim)
        
        return quantized, indices
```

Try testing it with your features to see the output shapes!

In [ ]:
class VectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x):
        batch, time, dim = x.shape
        x = x.view(batch, time, self.num_groups, -1)
        
        # Normalize for cosine similarity
        x_norm = torch.nn.functional.normalize(x, dim=-1)
        codebook_norm = torch.nn.functional.normalize(self.codebook, dim=-1)
        
        # Compute cosine similarity (use negative for "distance")
        similarity = torch.einsum('btgd,gvd->btgv', x_norm, codebook_norm)
        
        indices = similarity.argmax(dim=-1)  # (batch, time, num_groups)
        
        quantized = torch.stack([self.codebook[g, indices[:,:,g]] for g in range(self.num_groups)], dim=2)
        quantized = quantized.view(batch, time, dim)
        
        return quantized, indices

In [ ]:
quantizer=VectorQuantizer()
features_quant=quantizer(features)
features_quant

(tensor([[[-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687],
          [-1.0301,  0.6571,  2.0227,  ...,  0.6924, -1.3822,  0.7687],
          [-1.0301,  0.6571,  2.0227,  ...,  0.6924, -1.3822,  0.7687],
          ...,
          [ 0.2129,  0.7947,  0.1248,  ...,  0.6924, -1.3822,  0.7687],
          [-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687],
          [-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687]],
 
         [[-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687],
          [-1.0301,  0.6571,  2.0227,  ...,  0.6924, -1.3822,  0.7687],
          [-1.0301,  0.6571,  2.0227,  ...,  0.6924, -1.3822,  0.7687],
          ...,
          [ 0.2129,  0.7947,  0.1248,  ...,  0.6924, -1.3822,  0.7687],
          [-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687],
          [-0.3395,  0.6416,  0.2060,  ...,  0.6924, -1.3822,  0.7687]]],
        grad_fn=<ViewBackward0>),
 tensor([[[108,  65],
          [309,  65],
          [309,  65],
 

In [ ]:
quantizer_gumbel = GumbelVectorQuantizer()
features_quant_gumbel=quantizer_gumbel(features)
features_quant_gumbel

(tensor([[[ 0.7222, -0.0030, -0.2333,  ...,  0.7939,  0.3693,  0.3450],
          [ 0.3131, -1.6051,  0.7729,  ..., -1.1270, -1.1420,  0.2915],
          [ 0.2040,  0.4093,  1.2136,  ..., -0.1307,  1.9644,  1.1724],
          ...,
          [ 0.0724,  0.6067, -1.2861,  ...,  0.6907, -0.9655,  2.0006],
          [-1.6330,  0.3312, -1.4577,  ...,  0.7725,  1.5304, -0.7689],
          [ 0.8624,  1.1090,  0.1844,  ..., -0.0392,  0.2813, -1.5385]],
 
         [[ 0.8640, -0.6451, -1.4123,  ...,  0.6220, -0.0440, -0.3397],
          [-1.0296, -1.6321,  1.3314,  ..., -0.5549, -0.4859,  1.3474],
          [ 1.4479, -0.1359, -0.4299,  ..., -0.8750, -2.2273, -0.4456],
          ...,
          [ 0.3332,  1.6483, -0.9629,  ..., -1.5031, -0.7034,  0.1183],
          [-0.1885, -1.1305, -0.3464,  ...,  0.1562,  0.2245,  0.9386],
          [ 0.0835, -1.0238,  0.2639,  ..., -0.1307,  1.9644,  1.1724]]],
        grad_fn=<UnsafeViewBackward0>),
 tensor([[[[0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0.

like so?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! Both quantizers are working. You can see:

1. **VectorQuantizer**: Returns quantized features (batch=2, time=636, dim=512) and indices (batch=2, time=636, groups=2) showing which codebook entries were selected.

2. **GumbelVectorQuantizer**: Returns quantized features and the probability distributions (one-hot after gumbel_softmax with hard=True).

Now let's move to the **transformer encoder** - the context network that processes the masked features. This is where the model learns contextual representations.

Would you like me to provide a simplified transformer encoder module?

explain what is Gumbel doing?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great question! Let me break down what Gumbel softmax does:

**The Problem:** 
In standard quantization, you pick the closest codebook entry (argmax). But argmax isn't differentiable - you can't backpropagate through it to train the codebook.

**Gumbel Softmax Solution:**
It adds Gumbel noise to the logits (similarities) before applying softmax. This creates a differentiable approximation of sampling from a categorical distribution.

The key parameter is **temperature**:
- High temperature (e.g., 2.0): Soft, smooth distribution - more exploration
- Low temperature (e.g., 0.1): Sharp, nearly one-hot - closer to hard selection
- `hard=True`: During forward pass uses one-hot, but gradients flow as if it were soft

During training, you gradually decrease temperature (annealing) so it starts exploring different codebook entries, then converges to making hard choices.

Does this make sense? Would you like me to show what the temperature effect looks like with a simple example?

interesting, tell more, what does it mean add noise

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good question! The Gumbel noise is a specific type of random noise that helps with sampling.

**Adding noise means:**
When you have similarity scores for each codebook entry (e.g., [0.3, 0.5, 0.2]), instead of just picking the highest one deterministically, you add random Gumbel-distributed values to each score first.

So it becomes something like: [0.3 + noise₁, 0.5 + noise₂, 0.2 + noise₃]

Then you apply softmax with temperature. The noise makes it probabilistic - sometimes a lower-scoring entry might "win" due to lucky noise, which helps exploration during training.

**Why Gumbel specifically?**
Gumbel noise has a mathematical property: adding it and then taking argmax is equivalent to sampling from a categorical distribution. It's the "right" noise for this purpose.

In your code, `F.gumbel_softmax` does this automatically - it adds the Gumbel noise internally.

Does this clarify it, or would you like to see what happens with and without the noise?

why gumbel? what makes it specific

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The Gumbel distribution has a special mathematical property that makes it perfect for sampling:

**The Key Property:**
If you have probabilities [p₁, p₂, p₃, ...] and you want to sample from them, you can:
1. Add independent Gumbel noise to log(pᵢ) for each option
2. Take the argmax

This gives you **exactly** the same result as sampling from the categorical distribution with those probabilities!

**Why this matters:**
- Other noise distributions (like Gaussian) don't have this property
- Gumbel noise makes argmax equivalent to sampling, which is what we want
- The softmax version makes it differentiable while preserving this sampling behavior

It's named after Emil Julius Gumbel, who studied this distribution. The mathematical proof involves extreme value theory.

Does this help explain why Gumbel is the "right" choice here? Or would you like to move on to the transformer encoder?

so basically adding gumbel noise doesnt change the probs of categories is the property?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Not quite! Adding Gumbel noise **does** introduce randomness - it changes which category gets selected on each sample.

**The property is:**
- Over many samples, the **frequency** with which each category is selected matches the original probabilities
- It's a way to randomly sample that respects the probability distribution

**Example:**
If you have probabilities [0.7, 0.2, 0.1]:
- Without noise: Always pick the first one (boring, not differentiable)
- With Gumbel noise: Pick first ~70% of time, second ~20%, third ~10% (random, but correct distribution)

The magic is that `argmax(log(p) + Gumbel_noise)` is mathematically equivalent to sampling from the categorical distribution with probabilities p.

Does that distinction make sense? The noise adds randomness, but in a way that preserves the underlying probability distribution over many samples.

so if I run it long enough iterations then this keeps distrubtion equivalent?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Exactly! Over many iterations, the distribution of selected categories will converge to match the original probabilities.

**In practice for wav2vec2:**
- Each forward pass randomly selects codebook entries according to their similarity scores
- This randomness helps the model explore different quantizations during training
- As temperature decreases (annealing), it becomes more deterministic and settles on the best choices

The Gumbel noise ensures this exploration is "fair" - high-probability entries get picked more often, but low-probability ones still get a chance.

Ready to move on to the transformer encoder now?

## Transformer encoder

provide

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the transformer encoder:

```python
class TransformerEncoder(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, dropout=0.1):
        super().__init__()
        encoder_layer = nn.TransformerEncoderLayer(
            d_model=dim,
            nhead=num_heads,
            dim_feedforward=ff_dim,
            dropout=dropout,
            activation='gelu',
            batch_first=True
        )
        self.transformer = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
        
    def forward(self, x, mask=None):
        # x: (batch, time, dim)
        return self.transformer(x, src_key_padding_mask=mask)
```

Try creating it and passing your features through. What shape do you get?

In [ ]:
class TransformerEncoder(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, dropout=0.1):
        super().__init__()
        encoder_layer = nn.TransformerEncoderLayer(
            d_model=dim,
            nhead=num_heads,
            dim_feedforward=ff_dim,
            dropout=dropout,
            activation='gelu',
            batch_first=True
        )
        self.transformer = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
        
    def forward(self, x, mask=None):
        # x: (batch, time, dim)
        return self.transformer(x, src_key_padding_mask=mask)

does it have postional encoding

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

No, it doesn't! Good catch. Wav2vec2 uses **relative positional embeddings** (or convolutional positional embeddings) rather than the standard sinusoidal ones from the original Transformer.

For a simplified version, you could add:

1. **Convolutional positional embeddings** - a 1D conv layer that adds position info
2. **Learned absolute positional embeddings** - simpler, just add learned embeddings
3. **Skip it for now** - the CNN already provides some positional info through its receptive fields

Which approach would you like to use?

I want to use  relative positional embeddings

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great choice! Relative positional embeddings are what wav2vec2 actually uses.

There are different ways to implement them. The most common approach for wav2vec2 is to add relative position biases to the attention scores. This means modifying the attention mechanism to include a learned bias based on the distance between positions.

However, this requires customizing the attention layers inside the transformer. For a simpler approach that's still effective, we could use **convolutional positional embeddings** - a conv layer applied to the features before the transformer that encodes position information.

Which would you prefer:
1. Full relative position biases in attention (more complex, requires custom attention)
2. Convolutional positional embeddings (simpler, still effective)

Full relative position biases in attention (more complex, requires custom attention)

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! For full relative position biases, we need to create a custom attention mechanism that adds learned biases based on the relative distance between query and key positions.

The key idea: when position i attends to position j, we add a learned bias that depends on (i - j).

This requires:
1. A learnable relative position bias table (e.g., for distances from -max_distance to +max_distance)
2. Custom multi-head attention that adds these biases to attention scores before softmax
3. Replacing the standard TransformerEncoderLayer with our custom version

Since this is more involved, would you like me to provide the full custom attention code, or would you prefer to start with just the relative position bias computation and build up from there?

full custom codex

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the full custom attention with relative position biases:

```python
class RelativeMultiHeadAttention(nn.Module):
    def __init__(self, dim=512, num_heads=8, max_distance=128, dropout=0.1):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5
        
        self.qkv = nn.Linear(dim, dim * 3)
        self.out_proj = nn.Linear(dim, dim)
        self.dropout = nn.Dropout(dropout)
        
        # Relative position bias
        self.max_distance = max_distance
        self.rel_pos_bias = nn.Parameter(torch.randn(2 * max_distance + 1, num_heads))
        
    def get_relative_positions(self, seq_len):
        positions = torch.arange(seq_len, device=self.rel_pos_bias.device)
        relative_positions = positions[None, :] - positions[:, None]  # (seq_len, seq_len)
        # Clamp to max_distance
        relative_positions = torch.clamp(relative_positions, -self.max_distance, self.max_distance)
        # Shift to 0-indexed
        relative_positions = relative_positions + self.max_distance
        return relative_positions
    
    def forward(self, x, mask=None):
        batch, seq_len, dim = x.shape
        
        # QKV projection
        qkv = self.qkv(x).reshape(batch, seq_len, 3, self.num_heads, self.head_dim)
        q, k, v = qkv.permute(2, 0, 3, 1, 4)  # (batch, num_heads, seq_len, head_dim)
        
        # Attention scores
        attn = (q @ k.transpose(-2, -1)) * self.scale  # (batch, num_heads, seq_len, seq_len)
        
        # Add relative position bias
        rel_pos = self.get_relative_positions(seq_len)
        rel_bias = self.rel_pos_bias[rel_pos]  # (seq_len, seq_len, num_heads)
        attn = attn + rel_bias.permute(2, 0, 1)  # Add bias to attention scores
        
        if mask is not None:
            attn = attn.masked_fill(mask[:, None, None, :], float('-inf'))
        
        attn = attn.softmax(dim=-1)
        attn = self.dropout(attn)
        
        # Apply attention to values
        out = (attn @ v).transpose(1, 2).reshape(batch, seq_len, dim)
        out = self.out_proj(out)
        
        return out


class TransformerEncoderLayer(nn.Module):
    def __init__(self, dim=512, num_heads=8, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.self_attn = RelativeMultiHeadAttention(dim, num_heads, max_distance, dropout)
        self.ff = nn.Sequential(
            nn.Linear(dim, ff_dim),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(ff_dim, dim),
            nn.Dropout(dropout)
        )
        self.norm1 = nn.LayerNorm(dim)
        self.norm2 = nn.LayerNorm(dim)
        
    def forward(self, x, mask=None):
        x = x + self.self_attn(self.norm1(x), mask)
        x = x + self.ff(self.norm2(x))
        return x


class TransformerEncoder(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.layers = nn.ModuleList([
            TransformerEncoderLayer(dim, num_heads, ff_dim, dropout, max_distance)
            for _ in range(num_layers)
        ])
        
    def forward(self, x, mask=None):
        for layer in self.layers:
            x = layer(x, mask)
        return x
```

Try creating the new transformer and testing it with your features!

In [ ]:
class RelativeMultiHeadAttention(nn.Module):
    def __init__(self, dim=512, num_heads=8, max_distance=128, dropout=0.1):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5
        
        self.qkv = nn.Linear(dim, dim * 3)
        self.out_proj = nn.Linear(dim, dim)
        self.dropout = nn.Dropout(dropout)
        
        # Relative position bias
        self.max_distance = max_distance
        self.rel_pos_bias = nn.Parameter(torch.randn(2 * max_distance + 1, num_heads))
        
    def get_relative_positions(self, seq_len):
        positions = torch.arange(seq_len, device=self.rel_pos_bias.device)
        relative_positions = positions[None, :] - positions[:, None]  # (seq_len, seq_len)
        # Clamp to max_distance
        relative_positions = torch.clamp(relative_positions, -self.max_distance, self.max_distance)
        # Shift to 0-indexed
        relative_positions = relative_positions + self.max_distance
        return relative_positions
    
    def forward(self, x, mask=None):
        batch, seq_len, dim = x.shape
        
        # QKV projection
        qkv = self.qkv(x).reshape(batch, seq_len, 3, self.num_heads, self.head_dim)
        q, k, v = qkv.permute(2, 0, 3, 1, 4)  # (batch, num_heads, seq_len, head_dim)
        
        # Attention scores
        attn = (q @ k.transpose(-2, -1)) * self.scale  # (batch, num_heads, seq_len, seq_len)
        
        # Add relative position bias
        rel_pos = self.get_relative_positions(seq_len)
        rel_bias = self.rel_pos_bias[rel_pos]  # (seq_len, seq_len, num_heads)
        attn = attn + rel_bias.permute(2, 0, 1)  # Add bias to attention scores
        
        if mask is not None:
            attn = attn.masked_fill(mask[:, None, None, :], float('-inf'))
        
        attn = attn.softmax(dim=-1)
        attn = self.dropout(attn)
        
        # Apply attention to values
        out = (attn @ v).transpose(1, 2).reshape(batch, seq_len, dim)
        out = self.out_proj(out)
        
        return out


class TransformerEncoderLayer(nn.Module):
    def __init__(self, dim=512, num_heads=8, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.self_attn = RelativeMultiHeadAttention(dim, num_heads, max_distance, dropout)
        self.ff = nn.Sequential(
            nn.Linear(dim, ff_dim),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(ff_dim, dim),
            nn.Dropout(dropout)
        )
        self.norm1 = nn.LayerNorm(dim)
        self.norm2 = nn.LayerNorm(dim)
        
    def forward(self, x, mask=None):
        x = x + self.self_attn(self.norm1(x), mask)
        x = x + self.ff(self.norm2(x))
        return x


class TransformerEncoder(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.layers = nn.ModuleList([
            TransformerEncoderLayer(dim, num_heads, ff_dim, dropout, max_distance)
            for _ in range(num_layers)
        ])
        
    def forward(self, x, mask=None):
        for layer in self.layers:
            x = layer(x, mask)
        return x

In [ ]:
transformer=TransformerEncoder()
out=transformer(features_quant[0])

In [ ]:
out.shape

torch.Size([2, 636, 512])

like so?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! Your transformer with relative position biases is working. Input shape (2, 636, 512) → output shape (2, 636, 512).

Now you have all the core components! Next is **Phase 3: Pretraining** where we put everything together. We need:

1. A masking strategy (randomly mask some timesteps in the features)
2. A contrastive loss (predict the quantized targets for masked positions)
3. A training loop

The key idea: mask some feature timesteps, pass through transformer, then predict which quantized codes belong to the masked positions.

Should I provide the masking function first?

## Masking

provide masking code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the masking function:

```python
def mask_features(features, mask_prob=0.065, mask_length=10):
    """
    Mask spans of features for contrastive learning.
    
    Args:
        features: (batch, time, dim)
        mask_prob: probability of starting a mask span
        mask_length: length of each mask span
    
    Returns:
        masked_features: features with masks applied
        mask: boolean mask (True = masked position)
    """
    batch, time, dim = features.shape
    mask = torch.zeros(batch, time, dtype=torch.bool)
    
    for b in range(batch):
        # Sample mask start positions
        mask_starts = torch.rand(time) < mask_prob
        
        # Extend to spans
        for i in range(time):
            if mask_starts[i]:
                end = min(i + mask_length, time)
                mask[b, i:end] = True
    
    # Create masked features (replace with learned mask embedding or zeros)
    masked_features = features.clone()
    masked_features[mask] = 0  # Simple approach: zero out
    
    return masked_features, mask
```

Try it on your features and see what percentage gets masked!

In [ ]:
def mask_features(features, mask_prob=0.065, mask_length=10):
    """
    Mask spans of features for contrastive learning.
    
    Args:
        features: (batch, time, dim)
        mask_prob: probability of starting a mask span
        mask_length: length of each mask span
    
    Returns:
        masked_features: features with masks applied
        mask: boolean mask (True = masked position)
    """
    batch, time, dim = features.shape
    mask = torch.zeros(batch, time, dtype=torch.bool)
    
    for b in range(batch):
        # Sample mask start positions
        mask_starts = torch.rand(time) < mask_prob
        
        # Extend to spans
        for i in range(time):
            if mask_starts[i]:
                end = min(i + mask_length, time)
                mask[b, i:end] = True
    
    # Create masked features (replace with learned mask embedding or zeros)
    masked_features = features.clone()
    masked_features[mask] = 0  # Simple approach: zero out
    
    return masked_features, mask

In [ ]:
masked_features, mask=mask_features(features_quant_gumbel[0])
masked_features.shape, mask.shape

(torch.Size([2, 636, 512]), torch.Size([2, 636]))

like so?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, exactly! Your masking is working - same feature shape, plus a boolean mask showing which positions are masked.

You can check how much got masked with:
```python
mask.float().mean()
```

Now we need the **contrastive loss**. The idea: for masked positions, predict which quantized code they should map to. This is typically done with a projection head and InfoNCE loss.

Should I provide the loss function code?

In [ ]:
mask.float().mean()

tensor(0.4670)

## Contrastive loss

provide code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the contrastive loss implementation:

```python
class ContrastiveLoss(nn.Module):
    def __init__(self, dim=512, proj_dim=256, num_negatives=100, temperature=0.1):
        super().__init__()
        self.projection = nn.Linear(dim, proj_dim)
        self.num_negatives = num_negatives
        self.temperature = temperature
        
    def forward(self, context, targets, mask):
        """
        Args:
            context: transformer output (batch, time, dim)
            targets: quantized features (batch, time, dim)
            mask: boolean mask (batch, time)
        """
        # Project to lower dimension
        context_proj = self.projection(context[mask])  # (num_masked, proj_dim)
        target_proj = self.projection(targets[mask])   # (num_masked, proj_dim)
        
        # Normalize
        context_proj = F.normalize(context_proj, dim=-1)
        target_proj = F.normalize(target_proj, dim=-1)
        
        # Compute similarities (positive pairs)
        pos_sim = (context_proj * target_proj).sum(dim=-1) / self.temperature  # (num_masked,)
        
        # Sample negatives from other positions in batch
        batch_size = targets.shape[0] * targets.shape[1]
        all_targets = self.projection(targets.view(-1, targets.shape[-1]))
        all_targets = F.normalize(all_targets, dim=-1)
        
        # Random negative samples
        neg_indices = torch.randint(0, batch_size, (context_proj.shape[0], self.num_negatives))
        neg_targets = all_targets[neg_indices]  # (num_masked, num_negatives, proj_dim)
        
        # Negative similarities
        neg_sim = torch.bmm(neg_targets, context_proj.unsqueeze(-1)).squeeze(-1) / self.temperature
        
        # Contrastive loss (InfoNCE)
        logits = torch.cat([pos_sim.unsqueeze(1), neg_sim], dim=1)  # (num_masked, 1 + num_negatives)
        labels = torch.zeros(logits.shape[0], dtype=torch.long, device=logits.device)
        
        loss = F.cross_entropy(logits, labels)
        return loss
```

Try creating it and computing a loss with your masked features and transformer output!

In [ ]:
class ContrastiveLoss(nn.Module):
    def __init__(self, dim=512, proj_dim=256, num_negatives=100, temperature=0.1):
        super().__init__()
        self.projection = nn.Linear(dim, proj_dim)
        self.num_negatives = num_negatives
        self.temperature = temperature
        
    def forward(self, context, targets, mask):
        """
        Args:
            context: transformer output (batch, time, dim)
            targets: quantized features (batch, time, dim)
            mask: boolean mask (batch, time)
        """
        # Project to lower dimension
        context_proj = self.projection(context[mask])  # (num_masked, proj_dim)
        target_proj = self.projection(targets[mask])   # (num_masked, proj_dim)
        
        # Normalize
        context_proj = F.normalize(context_proj, dim=-1)
        target_proj = F.normalize(target_proj, dim=-1)
        
        # Compute similarities (positive pairs)
        pos_sim = (context_proj * target_proj).sum(dim=-1) / self.temperature  # (num_masked,)
        
        # Sample negatives from other positions in batch
        batch_size = targets.shape[0] * targets.shape[1]
        all_targets = self.projection(targets.view(-1, targets.shape[-1]))
        all_targets = F.normalize(all_targets, dim=-1)
        
        # Random negative samples
        neg_indices = torch.randint(0, batch_size, (context_proj.shape[0], self.num_negatives))
        neg_targets = all_targets[neg_indices]  # (num_masked, num_negatives, proj_dim)
        
        # Negative similarities
        neg_sim = torch.bmm(neg_targets, context_proj.unsqueeze(-1)).squeeze(-1) / self.temperature
        
        # Contrastive loss (InfoNCE)
        logits = torch.cat([pos_sim.unsqueeze(1), neg_sim], dim=1)  # (num_masked, 1 + num_negatives)
        labels = torch.zeros(logits.shape[0], dtype=torch.long, device=logits.device)
        
        loss = F.cross_entropy(logits, labels)
        return loss

In [ ]:
loss_fun=ContrastiveLoss()
loss = loss_fun(out, features_quant_gumbel[0], mask)
loss

tensor(4.8583, grad_fn=<NllLossBackward0>)

what next?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great! Your loss is computed successfully. Now let's put everything together into a **Wav2Vec2 model** class that combines all components, then create a training loop.

Would you like me to provide:
1. The full Wav2Vec2 model class that wraps all components
2. Then the pretraining loop

Or would you prefer to try building the model class yourself first?

## Full model

In [1]:
import torch
from torch.utils.data import Dataset, DataLoader
import torchaudio
from torch import nn
import torch.nn.functional as F

class AudioDataset(Dataset):
    def __init__(self, audio_files, sample_rate=16000):
        self.audio_files = audio_files
        self.sample_rate = sample_rate
    
    def __len__(self):
        return len(self.audio_files)
    
    def __getitem__(self, idx):
        waveform, sr = torchaudio.load(self.audio_files[idx])
        # Resample if needed
        if sr != self.sample_rate:
            resampler = torchaudio.transforms.Resample(sr, self.sample_rate)
            waveform = resampler(waveform)
        # Convert to mono if stereo
        if waveform.shape[0] > 1:
            waveform = torch.mean(waveform, dim=0, keepdim=True)
        return waveform.squeeze(0)

def collate_fn(batch):
    # Pad sequences to same length
    lengths = [x.shape[0] for x in batch]
    max_len = max(lengths)
    padded = torch.stack([torch.nn.functional.pad(x, (0, max_len - x.shape[0])) for x in batch])
    return padded, torch.tensor(lengths)


In [2]:
class FeatureExtractor(nn.Module):
    def __init__(self, conv_layers=[(512, 10, 5)] + [(512, 3, 2)] * 4 + [(512, 2, 2)] * 2):
        super().__init__()
        layers = []
        in_channels = 1
        for out_channels, kernel_size, stride in conv_layers:
            layers.extend([
                nn.Conv1d(in_channels, out_channels, kernel_size, stride, bias=False),
                nn.GroupNorm(out_channels, out_channels),
                nn.GELU()
            ])
            in_channels = out_channels
        self.conv = nn.Sequential(*layers)
    
    def forward(self, x):
        x = x.unsqueeze(1)
        x = self.conv(x)
        return x.transpose(1, 2)

class GumbelVectorQuantizer(nn.Module):
    def __init__(self, num_vars=320, num_groups=2, dim=512, temp=(2.0, 0.5, 0.999995)):
        super().__init__()
        self.num_vars = num_vars
        self.num_groups = num_groups
        self.dim = dim
        
        self.max_temp, self.min_temp, self.temp_decay = temp
        self.curr_temp = self.max_temp
        
        self.codebook = nn.Parameter(torch.randn(num_groups, num_vars, dim // num_groups))
        
    def forward(self, x, training=True):
        batch, time, dim = x.shape
        x = x.view(batch, time, self.num_groups, -1)
        
        x_norm = F.normalize(x, dim=-1)
        codebook_norm = F.normalize(self.codebook, dim=-1)
        
        sim = torch.einsum('btgd,gvd->btgv', x_norm, codebook_norm)
        sim = sim.reshape(batch * time * self.num_groups, self.num_vars)
        
        if training:
            probs = F.gumbel_softmax(sim / self.curr_temp, hard=True, dim=-1)
        else:
            indices = sim.argmax(dim=-1)
            probs = F.one_hot(indices, self.num_vars).float()
        
        probs = probs.view(batch * time, self.num_groups, self.num_vars)
        quantized = torch.einsum('bgv,gvd->bgd', probs, self.codebook)
        quantized = quantized.reshape(batch, time, dim)
        
        return quantized, probs.view(batch, time, self.num_groups, self.num_vars)

In [3]:
class RelativeMultiHeadAttention(nn.Module):
    def __init__(self, dim=512, num_heads=8, max_distance=128, dropout=0.1):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5
        
        self.qkv = nn.Linear(dim, dim * 3)
        self.out_proj = nn.Linear(dim, dim)
        self.dropout = nn.Dropout(dropout)
        
        self.max_distance = max_distance
        self.rel_pos_bias = nn.Parameter(torch.randn(2 * max_distance + 1, num_heads))
        
    def get_relative_positions(self, seq_len):
        positions = torch.arange(seq_len, device=self.rel_pos_bias.device)
        relative_positions = positions[None, :] - positions[:, None]
        relative_positions = torch.clamp(relative_positions, -self.max_distance, self.max_distance)
        relative_positions = relative_positions + self.max_distance
        return relative_positions
    
    def forward(self, x, mask=None):
        batch, seq_len, dim = x.shape
        
        qkv = self.qkv(x).reshape(batch, seq_len, 3, self.num_heads, self.head_dim)
        q, k, v = qkv.permute(2, 0, 3, 1, 4)
        
        attn = (q @ k.transpose(-2, -1)) * self.scale
        
        rel_pos = self.get_relative_positions(seq_len)
        rel_bias = self.rel_pos_bias[rel_pos]
        attn = attn + rel_bias.permute(2, 0, 1)
        
        if mask is not None:
            attn = attn.masked_fill(mask[:, None, None, :], float('-inf'))
        
        attn = attn.softmax(dim=-1)
        attn = self.dropout(attn)
        
        out = (attn @ v).transpose(1, 2).reshape(batch, seq_len, dim)
        out = self.out_proj(out)
        
        return out

class TransformerEncoderLayer(nn.Module):
    def __init__(self, dim=512, num_heads=8, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.self_attn = RelativeMultiHeadAttention(dim, num_heads, max_distance, dropout)
        self.ff = nn.Sequential(
            nn.Linear(dim, ff_dim),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(ff_dim, dim),
            nn.Dropout(dropout)
        )
        self.norm1 = nn.LayerNorm(dim)
        self.norm2 = nn.LayerNorm(dim)
        
    def forward(self, x, mask=None):
        x = x + self.self_attn(self.norm1(x), mask)
        x = x + self.ff(self.norm2(x))
        return x

class TransformerEncoder(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, dropout=0.1, max_distance=128):
        super().__init__()
        self.layers = nn.ModuleList([
            TransformerEncoderLayer(dim, num_heads, ff_dim, dropout, max_distance)
            for _ in range(num_layers)
        ])
        
    def forward(self, x, mask=None):
        for layer in self.layers:
            x = layer(x, mask)
        return x

In [4]:
def mask_features(features, mask_prob=0.065, mask_length=10):
    batch, time, dim = features.shape
    mask = torch.zeros(batch, time, dtype=torch.bool)
    
    for b in range(batch):
        mask_starts = torch.rand(time) < mask_prob
        for i in range(time):
            if mask_starts[i]:
                end = min(i + mask_length, time)
                mask[b, i:end] = True
    
    masked_features = features.clone()
    masked_features[mask] = 0
    
    return masked_features, mask

class ContrastiveLoss(nn.Module):
    def __init__(self, dim=512, proj_dim=256, num_negatives=100, temperature=0.1):
        super().__init__()
        self.projection = nn.Linear(dim, proj_dim)
        self.num_negatives = num_negatives
        self.temperature = temperature
        
    def forward(self, context, targets, mask):
        context_proj = self.projection(context[mask])
        target_proj = self.projection(targets[mask])
        
        context_proj = F.normalize(context_proj, dim=-1)
        target_proj = F.normalize(target_proj, dim=-1)
        
        pos_sim = (context_proj * target_proj).sum(dim=-1) / self.temperature
        
        batch_size = targets.shape[0] * targets.shape[1]
        all_targets = self.projection(targets.view(-1, targets.shape[-1]))
        all_targets = F.normalize(all_targets, dim=-1)
        
        neg_indices = torch.randint(0, batch_size, (context_proj.shape[0], self.num_negatives))
        neg_targets = all_targets[neg_indices]
        
        neg_sim = torch.bmm(neg_targets, context_proj.unsqueeze(-1)).squeeze(-1) / self.temperature
        
        logits = torch.cat([pos_sim.unsqueeze(1), neg_sim], dim=1)
        labels = torch.zeros(logits.shape[0], dtype=torch.long, device=logits.device)
        
        loss = F.cross_entropy(logits, labels)
        return loss

In [5]:
class Wav2Vec2(nn.Module):
    def __init__(self, dim=512, num_heads=8, num_layers=4, ff_dim=2048, 
                 num_vars=320, num_groups=2, proj_dim=256):
        super().__init__()
        self.feature_extractor = FeatureExtractor()
        self.quantizer = GumbelVectorQuantizer(num_vars, num_groups, dim)
        self.transformer = TransformerEncoder(dim, num_heads, num_layers, ff_dim)
        self.loss_fn = ContrastiveLoss(dim, proj_dim)
        
    def forward(self, waveform, mask_prob=0.065, mask_length=10):
        # Extract features
        features = self.feature_extractor(waveform)
        
        # Quantize (for targets)
        with torch.no_grad():
            quantized, _ = self.quantizer(features, training=False)
        
        # Mask features
        masked_features, mask = mask_features(features, mask_prob, mask_length)
        
        # Context network
        context = self.transformer(masked_features)
        
        # Compute loss
        loss = self.loss_fn(context, quantized, mask)
        
        return loss, context

In [6]:

dataset = AudioDataset(["emt16k.wav", "emt16k.wav", "emt16k.wav"])
dataloader = DataLoader(dataset, batch_size=2, collate_fn=collate_fn, shuffle=True)
batch=next(iter(dataloader))

In [7]:
model=Wav2Vec2()
model_output=model(batch[0])

In [8]:
model_output

(tensor(4.9038, grad_fn=<NllLossBackward0>),
 tensor([[[-0.8654,  1.0914, -0.5556,  ..., -0.2947, -0.1307, -0.4022],
          [-0.3746,  0.6515, -0.1099,  ..., -0.0615, -0.2931, -0.5915],
          [-0.3980, -0.2671, -0.5916,  ..., -0.0031,  0.1800, -0.0125],
          ...,
          [-0.3713,  0.6989, -0.2076,  ...,  0.3091, -0.0201, -0.6790],
          [-0.5497,  1.3568, -0.4534,  ...,  0.1363, -0.1382, -0.6732],
          [-0.7374,  1.0159, -0.6222,  ...,  0.0351, -0.3687, -0.5434]],
 
         [[-0.3105,  0.8129, -0.7929,  ...,  0.0384,  0.2373, -0.0039],
          [-0.2498,  0.3805, -0.0323,  ...,  0.1770, -0.1055, -0.3956],
          [-0.1047,  0.5666, -0.3326,  ..., -0.1909, -0.2764, -0.5848],
          ...,
          [-0.6060,  0.4783, -0.5618,  ..., -0.0202, -0.1729, -0.4106],
          [-0.3155,  0.4133, -0.2951,  ...,  0.0693, -0.0417, -0.3673],
          [-0.7998, -0.1483, -0.5607,  ...,  0.4085,  0.1419, -0.4339]]],
        grad_fn=<AddBackward0>))

## Pretraining

provide pretraining code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the pretraining loop:

```python
def pretrain(model, dataloader, num_epochs=10, lr=1e-4):
    optimizer = torch.optim.AdamW(model.parameters(), lr=lr)
    model.train()
    
    for epoch in range(num_epochs):
        total_loss = 0
        for batch_idx, (waveforms, lengths) in enumerate(dataloader):
            optimizer.zero_grad()
            
            loss, _ = model(waveforms)
            loss.backward()
            optimizer.step()
            
            # Temperature annealing for Gumbel
            model.quantizer.curr_temp = max(
                model.quantizer.max_temp * model.quantizer.temp_decay,
                model.quantizer.min_temp
            )
            
            total_loss += loss.item()
            
            if batch_idx % 10 == 0:
                print(f"Epoch {epoch}, Batch {batch_idx}, Loss: {loss.item():.4f}, Temp: {model.quantizer.curr_temp:.4f}")
        
        avg_loss = total_loss / len(dataloader)
        print(f"Epoch {epoch} complete. Average Loss: {avg_loss:.4f}")
    
    return model
```

Try running it with your model and dataloader. How many epochs do you want to train for?

In [9]:
num_epochs=3
lr=1e-4
optimizer = torch.optim.AdamW(model.parameters(), lr=lr)
model.train()

for epoch in range(num_epochs):
    total_loss = 0
    for batch_idx, (waveforms, lengths) in enumerate(dataloader):
        optimizer.zero_grad()
        
        loss, _ = model(waveforms)
        loss.backward()
        optimizer.step()
        
        # Temperature annealing for Gumbel
        model.quantizer.curr_temp = max(
            model.quantizer.max_temp * model.quantizer.temp_decay,
            model.quantizer.min_temp
        )
        
        total_loss += loss.item()
        
        if batch_idx % 10 == 0:
            print(f"Epoch {epoch}, Batch {batch_idx}, Loss: {loss.item():.4f}, Temp: {model.quantizer.curr_temp:.4f}")
    
    avg_loss = total_loss / len(dataloader)
    print(f"Epoch {epoch} complete. Average Loss: {avg_loss:.4f}")

Epoch 0, Batch 0, Loss: 4.9062, Temp: 2.0000
Epoch 0 complete. Average Loss: 4.9213
Epoch 1, Batch 0, Loss: 5.1381, Temp: 2.0000
Epoch 1 complete. Average Loss: 5.0442
Epoch 2, Batch 0, Loss: 4.7615, Temp: 2.0000
Epoch 2 complete. Average Loss: 4.7410


## Finetuning

provide example how to finetune for ASR this model

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

For ASR finetuning, you need to add a classification head on top of the pretrained model and train it with labeled transcriptions.

Here's the basic structure:

```python
class Wav2Vec2ForASR(nn.Module):
    def __init__(self, pretrained_model, vocab_size):
        super().__init__()
        self.wav2vec2 = pretrained_model
        self.lm_head = nn.Linear(512, vocab_size)
        
    def forward(self, waveform):
        # Freeze or unfreeze wav2vec2 layers as needed
        features = self.wav2vec2.feature_extractor(waveform)
        context = self.wav2vec2.transformer(features)
        logits = self.lm_head(context)
        return logits  # (batch, time, vocab_size)
```

For the training loop, you'll need:
1. A dataset with audio files and their transcriptions
2. CTC loss (Connectionist Temporal Classification) to handle alignment

Do you have labeled transcriptions for your audio, or should I show you how to create a simple synthetic example for demonstration?

In [10]:
class Wav2Vec2ForASR(nn.Module):
    def __init__(self, pretrained_model, vocab_size):
        super().__init__()
        self.wav2vec2 = pretrained_model
        self.lm_head = nn.Linear(512, vocab_size)
        
    def forward(self, waveform):
        # Freeze or unfreeze wav2vec2 layers as needed
        features = self.wav2vec2.feature_extractor(waveform)
        context = self.wav2vec2.transformer(features)
        logits = self.lm_head(context)
        return logits  # (batch, time, vocab_size)

how to make finetuning with toy examples

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

For toy finetuning, you can create synthetic labels. Here's a simple approach:

1. **Create a small vocabulary** (e.g., 30 characters: a-z, space, punctuation)
2. **Generate fake transcriptions** for your audio file
3. **Use CTC loss** which handles the alignment between audio frames and text

Here's the setup:

```python
import torch.nn as nn

# Vocabulary
vocab = list("abcdefghijklmnopqrstuvwxyz .,")  # 30 chars
vocab_size = len(vocab)
char_to_idx = {c: i for i, c in enumerate(vocab)}

# Fake transcription for your audio
transcription = "hello world"
```

For CTC loss, you need:
- Input: logits from model (batch, time, vocab_size)
- Target: encoded transcription (batch, target_length)
- Input lengths and target lengths

Would you like me to provide the complete finetuning code with CTC loss, or would you prefer to try implementing it yourself with guidance?

provide code

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's the complete finetuning code with CTC loss:

```python
# Vocabulary setup
vocab = list("abcdefghijklmnopqrstuvwxyz .,")
vocab_size = len(vocab)
char_to_idx = {c: i for i, c in enumerate(vocab)}

# Create ASR model
asr_model = Wav2Vec2ForASR(model, vocab_size)

# Fake dataset with transcriptions
class ASRDataset(Dataset):
    def __init__(self, audio_files, transcriptions):
        self.audio_files = audio_files
        self.transcriptions = transcriptions
        
    def __len__(self):
        return len(self.audio_files)
    
    def __getitem__(self, idx):
        waveform, sr = torchaudio.load(self.audio_files[idx])
        if waveform.shape[0] > 1:
            waveform = torch.mean(waveform, dim=0)
        else:
            waveform = waveform.squeeze(0)
        
        # Encode transcription
        text = self.transcriptions[idx]
        encoded = torch.tensor([char_to_idx[c] for c in text if c in char_to_idx])
        
        return waveform, encoded

def asr_collate_fn(batch):
    waveforms, transcriptions = zip(*batch)
    
    # Pad waveforms
    wav_lengths = [w.shape[0] for w in waveforms]
    max_wav_len = max(wav_lengths)
    padded_wavs = torch.stack([F.pad(w, (0, max_wav_len - w.shape[0])) for w in waveforms])
    
    # Pad transcriptions
    text_lengths = [t.shape[0] for t in transcriptions]
    max_text_len = max(text_lengths)
    padded_texts = torch.stack([F.pad(t, (0, max_text_len - t.shape[0])) for t in transcriptions])
    
    return padded_wavs, torch.tensor(wav_lengths), padded_texts, torch.tensor(text_lengths)

# Training function
def finetune_asr(model, dataloader, num_epochs=5, lr=1e-4):
    optimizer = torch.optim.AdamW(model.parameters(), lr=lr)
    ctc_loss = nn.CTCLoss(blank=0, zero_infinity=True)
    model.train()
    
    for epoch in range(num_epochs):
        total_loss = 0
        for batch_idx, (waveforms, wav_lens, targets, target_lens) in enumerate(dataloader):
            optimizer.zero_grad()
            
            logits = model(waveforms)  # (batch, time, vocab)
            log_probs = F.log_softmax(logits, dim=-1)
            
            # CTC expects (time, batch, vocab)
            log_probs = log_probs.transpose(0, 1)
            
            # Input lengths (after feature extraction downsampling)
            input_lengths = wav_lens // 320  # Approximate downsampling factor
            
            loss = ctc_loss(log_probs, targets, input_lengths, target_lens)
            loss.backward()
            optimizer.step()
            
            total_loss += loss.item()
            
            if batch_idx % 5 == 0:
                print(f"Epoch {epoch}, Batch {batch_idx}, Loss: {loss.item():.4f}")
        
        avg_loss = total_loss / len(dataloader)
        print(f"Epoch {epoch} complete. Average Loss: {avg_loss:.4f}")
    
    return model

# Create toy dataset
asr_dataset = ASRDataset(
    ["emt16k.wav"] * 3,
    ["hello world", "hello world", "hello world"]
)
asr_dataloader = DataLoader(asr_dataset, batch_size=2, collate_fn=asr_collate_fn)

# Finetune
asr_model = finetune_asr(asr_model, asr_dataloader, num_epochs=3)
```

Try running this to see the ASR finetuning in action!

In [11]:
# Vocabulary setup
vocab = list("abcdefghijklmnopqrstuvwxyz .,")
vocab_size = len(vocab)
char_to_idx = {c: i for i, c in enumerate(vocab)}

# Create ASR model
asr_model = Wav2Vec2ForASR(model, vocab_size)

# Fake dataset with transcriptions
class ASRDataset(Dataset):
    def __init__(self, audio_files, transcriptions):
        self.audio_files = audio_files
        self.transcriptions = transcriptions
        
    def __len__(self):
        return len(self.audio_files)
    
    def __getitem__(self, idx):
        waveform, sr = torchaudio.load(self.audio_files[idx])
        if waveform.shape[0] > 1:
            waveform = torch.mean(waveform, dim=0)
        else:
            waveform = waveform.squeeze(0)
        
        # Encode transcription
        text = self.transcriptions[idx]
        encoded = torch.tensor([char_to_idx[c] for c in text if c in char_to_idx])
        
        return waveform, encoded

def asr_collate_fn(batch):
    waveforms, transcriptions = zip(*batch)
    
    # Pad waveforms
    wav_lengths = [w.shape[0] for w in waveforms]
    max_wav_len = max(wav_lengths)
    padded_wavs = torch.stack([F.pad(w, (0, max_wav_len - w.shape[0])) for w in waveforms])
    
    # Pad transcriptions
    text_lengths = [t.shape[0] for t in transcriptions]
    max_text_len = max(text_lengths)
    padded_texts = torch.stack([F.pad(t, (0, max_text_len - t.shape[0])) for t in transcriptions])
    
    return padded_wavs, torch.tensor(wav_lengths), padded_texts, torch.tensor(text_lengths)

# Training function
def finetune_asr(model, dataloader, num_epochs=2, lr=1e-4):
    optimizer = torch.optim.AdamW(model.parameters(), lr=lr)
    ctc_loss = nn.CTCLoss(blank=0, zero_infinity=True)
    model.train()
    
    for epoch in range(num_epochs):
        total_loss = 0
        for batch_idx, (waveforms, wav_lens, targets, target_lens) in enumerate(dataloader):
            optimizer.zero_grad()
            
            logits = model(waveforms)  # (batch, time, vocab)
            log_probs = F.log_softmax(logits, dim=-1)
            
            # CTC expects (time, batch, vocab)
            log_probs = log_probs.transpose(0, 1)
            
            # Input lengths (after feature extraction downsampling)
            input_lengths = wav_lens // 320  # Approximate downsampling factor
            
            loss = ctc_loss(log_probs, targets, input_lengths, target_lens)
            loss.backward()
            optimizer.step()
            
            total_loss += loss.item()
            
            if batch_idx % 5 == 0:
                print(f"Epoch {epoch}, Batch {batch_idx}, Loss: {loss.item():.4f}")
        
        avg_loss = total_loss / len(dataloader)
        print(f"Epoch {epoch} complete. Average Loss: {avg_loss:.4f}")
    
    return model

# Create toy dataset
asr_dataset = ASRDataset(
    ["emt16k.wav"] * 3,
    ["hello world", "hello world", "hello world"]
)
asr_dataloader = DataLoader(asr_dataset, batch_size=2, collate_fn=asr_collate_fn)

# Finetune
asr_model = finetune_asr(asr_model, asr_dataloader, num_epochs=3)

Epoch 0, Batch 0, Loss: 125.8491
Epoch 0 complete. Average Loss: 68.4222
Epoch 1, Batch 0, Loss: 9.7192
Epoch 1 complete. Average Loss: 10.1696
Epoch 2, Batch 0, Loss: 10.7034
Epoch 2 complete. Average Loss: 10.1852
